server/src/handlers/influxdb.rs

```diff
@@ -71,6 +77,303 @@ impl From<Bytes> for WriteRequest {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut one_series = serializer.serialize_map(Some(4))?;
+        one_series.serialize_entry("name", &self.name)?;
+        if let Some(tags) = &self.tags {
+            one_series.serialize_entry("tags", &tags)?;
+        }
+        one_series.serialize_entry("columns", &self.columns)?;
+        one_series.serialize_entry("values", &self.values)?;
+
+        one_series.end()
+    }
+}
+
+#[derive(Debug)]
+struct Tags(Vec<(String, String)>);
+
+impl Serialize for Tags {
+    fn serialize<S>(&self, serializer: S) -> std::result::Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let mut tags = serializer.serialize_map(Some(self.0.len()))?;
+
+        for (tagk, tagv) in &self.0 {
+            tags.serialize_entry(tagk, tagv)?;
+        }
+
+        tags.end()
+    }
+}
+
+/// [InfluxqlResult] builder
+#[derive(Default)]
+pub struct InfluxqlResultBuilder {
+    /// Id of the related influxql
+    statement_id: u32,
+
+    /// Schema of influxql query result
+    ///
+    /// Its format is like:
+    /// measurement |
+    /// tag_1..tag_n(columns in `group by`) |
+    /// time |
+    /// column_1..column_n(column in `projection` but not in `group by`)
+    column_schemas: Vec<ColumnSchema>,
+
+    /// Tags part in schema
+    group_by_tag_col_idxs: Vec<usize>,
+
+    /// Columns part in schema(include `time`)
+    value_col_idxs: Vec<usize>,
+
+    /// Mapping group key(`measurement` + `tag values`) to column values,
+    ///
+    /// NOTE: because tag keys in `group by` clause are same in each sub result,
+    /// we just use the `measurement` + `tag values` to distinguish them.
+    group_key_to_idx: HashMap<GroupKey, usize>,
+
+    /// Column values
+    value_groups: Vec<Vec<Vec<Datum>>>,
+}
+
+impl InfluxqlResultBuilder {
+    pub fn new(record_schema: &RecordSchema, statement_id: u32) -> Result<Self> {
+        let column_schemas = record_schema.columns().to_owned();
+        ensure!(
+            !column_schemas.is_empty(),
+            InfluxDbHandlerNoCause {
+                msg: "empty schema",
+            }
+        );
+
+        // Find the tags part and columns part from schema.
+        let mut group_by_col_idxs = Vec::new();
+        let mut value_col_idxs = Vec::new();
+
+        let mut col_iter = column_schemas.iter().enumerate();
+        // The first column may be measurement column in normal.
+        ensure!(col_iter.next().unwrap().1.name == CERESDB_MEASUREMENT_COLUMN_NAME, InfluxDbHandlerNoCause {
+            msg: format!("invalid schema whose first column is not measurement column, schema:{column_schemas:?}"),
+        });
+
+        // The group by tags will be placed after measurement and before time column.
+        let mut searching_group_by_tags = true;
+        for (idx, col) in col_iter {
+            if col.data_type.is_timestamp() {
```

Comment:
What if there are two timestamp columns?

Reply:
It will ensure the projection is like that:
```
measurement  | group by tag | timestamp key
```
we just need to find the first timestamp column(timestamp key).